plugins/Goals/PagesBeforeCalculator.php

```diff
@@ -0,0 +1,154 @@
+        } else {
+            // Specific sites
+            $sites = explode(',', $idSite);
+        }
+
+        if ($lastN) {
+
+            // Since MySQL doesn't support multi-table updates with a LIMIT clause we will find the exact date time of
+            // the lastN record and use that as a date range start with the current date time as the date range end
+            $sql = "
+                    SELECT MIN(c.server_time) 
+                    FROM " . Common::prefixTable('log_conversion') . " c                                 
+                    ";
+
+            $where = '';
+             if (!$forceRecalc) {
+                 $where .= " AND c.pageviews_before IS NULL";
+             }
+
+            $bind = [];
+            if ($idGoal !== null) {
+                $where .= ' AND c.idgoal = ? ';
+                $bind[] = $idGoal;
+            }
+            if ($idSite !== null) {
+                $where .= ' AND c.idsite = ? ';
+                $bind[] = $idSite;
+            }
+
+            if ($where !== '') {
+                $sql .= ' WHERE '.ltrim($where, 'AND ');
+            }
+
+            $sql .= " ORDER BY c.server_time DESC LIMIT " . $lastN;
+
+            $result = Db::fetchOne($sql, $bind);
+
+            $startDatetime = $result;
+            $endDatetime = Date::factory('now')->getDatetime();
+        }
+
+        foreach ($sites as $site) {
+
+            if ($idGoal === null) {
+                // All goals
+                $goalsModel = new GoalsModel();
+                $goals = array_column($goalsModel->getActiveGoals([$site]), 'idgoal');
+
+                // Include ecommerce orders if enabled for the site
+                if (Site::isEcommerceEnabledFor($site)) {
+                    $goals[] = GoalManager::IDGOAL_ORDER;
+                }
+            } else {
+                // Specific goals
+                $goals = explode(',', $idGoal);
+            }
+
+            foreach ($goals as $goal) {
+
+                $sql = "
+                UPDATE " . Common::prefixTable('log_conversion') . " c
+                LEFT JOIN (
+                    SELECT COUNT(va.idvisit) AS pages_before, va.idvisit, va.server_time
+                    FROM " . Common::prefixTable('log_link_visit_action') . " va
+                    LEFT JOIN " . Common::prefixTable('log_action') . " a ON a.idaction = va.idaction_url AND a.type = 1
+                    GROUP BY va.idvisit
+                ) AS a ON a.idvisit = c.idvisit AND a.server_time <= c.server_time
+                SET c.pageviews_before = a.pages_before
+                WHERE c.idsite = ? AND c.idgoal = ?                    
+                ";
+
+                if (!$forceRecalc) {
+                     $sql .= " AND c.pageviews_before IS NULL";
+                }
+
+                $bind = [$site, $goal];
+
+                if (!empty($startDatetime)) {
+                    $sql .= " AND c.server_time >= ?";
+                    $bind[] = $startDatetime;
```

Comment:
Did you consider timezones? When looking at this again I was wondering if that would work for sites that use another timezone than UTC. All data in the database is stored in UTC, but when a site is archived it might look at times that are before or after the UTC time. So guess we might need to apply something like this to the start and endtime:

```php
$timezone = Site::getTimezoneFor($idsite);
$bind[] = Date::factory($startDatetime, $timezone)->getDateTime();
```

Reply:
Good point. The migration update for the switchover is for today and yesterday which is enough to cover any timezone differences, but standalone use of the command should adjust the timezone :+1: